### DISCUSSION TITLE
CRASH: Could not cast value of type 'Observation.ObservationRegistrar' to 'Perception._PerceptionRegistrar'

### DISCUSSION THREAD

**Discussion by iharandreyev** [pointfreeco/swift-composable-architecture#3113] [General]
### Description

The error in question happens when I'm trying to test the following feature

```
import ComposableArchitecture
import Foundation

@ObservableState
enum RemoteResourceState<Value> {
    case none
    case loading
    case loaded(Value)
    case failed(Error)
}

extension RemoteResourceState: Equatable where Value: Equatable {
    static func == (lhs: Self, rhs: Self) -> Bool {
        switch (lhs, rhs) {
        case (.none, .none): return true
        case (.loading, .loading): return true
        case let (.loaded(lhs), .loaded(rhs)):
            return lhs == rhs
        case let (.failed(lhs), .failed(rhs)):
            return "\(lhs)" == "\(rhs)"
        default:
            return false
        }
    }
}

enum RemoteResourceAction<Value> {
    case load
    case succeed(Value)
    case fail(Error)
}

extension RemoteResourceAction: Equatable where Value: Equatable {
    static func == (lhs: Self, rhs: Self) -> Bool {
        switch (lhs, rhs) {
        case (.load, .load): return true
        case let (.succeed(lhs), .succeed(rhs)):
            return lhs == rhs
        case let (.fail(lhs), .fail(rhs)):
            return "\(lhs)" == "\(rhs)"
        default:
            return false
        }
    }
}

struct RemoteResource<Value>: Reducer {
    typealias State = RemoteResourceState<Value>
    typealias Action = RemoteResourceAction<Value>
    
    var body: some ReducerOf<Self> {
        EmptyReducer()
    }
}
```

with a test below

```
import XCTest
import ComposableArchitecture
@testable import MyApp

final class RemoteResourceTests: XCTestCase {
    func testLoadFirst() async {
        typealias SUT = RemoteResource<TestResource>
        
        let store = TestStoreOf<SUT>(
            initialState: .none,
            reducer: SUT.init
        )
        
        await store.send(.load) { state in
            state = .loading
        }
    }
}

// MARK: - Helpers

private struct TestResource: Identifiable, Equatable {
    let id: Int = UUID().hashValue
}
```

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [ ] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

Test fails with an invalid behavior error

### Actual behavior

Test crashes with `Could not cast value of type 'Observation.ObservationRegistrar' to 'Perception._PerceptionRegistrar'` error in `Perception/PerceptionRegistrar.swift:52`

### Steps to reproduce

Run the code in the description

### The Composable Architecture version information

1.10.4

### Destination operating system

iOS 15.0

### Xcode version information

15.4

### Swift Compiler version information

```shell
swift-driver version: 1.90.11.1 Apple Swift version 5.10 (swiftlang-5.10.0.13 clang-1500.3.9.4)
Target: arm64-apple-macosx14.0
```

**Comment by iharandreyev** [comment]
The issue occurs when I use iOS 17.5 simulator as a test run destination.
When running on iOS 15.5 simulator, there's no crash

**Comment by mbrandonw** [comment]
Hi @iharandreyev, I am not able to reproduce this. I have run it in a variety of simulators and it all seems to work fine. Can you please provide a project that demonstrates the problem?

I am also going to convert this to a discussion for the time being because it's not clear there is an actual issue with the library. Let's continue the conversation over there.

**Comment by iharandreyev** [comment]
@mbrandonw Here is a [link](https://github.com/iharandreyev/Tuist-CTA-LinkingIssue) to a repo where the issue is reproduced.

I now think it may have to do something with the fact I'm using tuist to generate the project.
I had some linking issues (as the repo name suggests), but after updating tuist to 4.14.0, linking issues went away.
However, I do get a weird warning

```
The following warnings need attention:
 · Invalid product name 'swift-composable-architecture_ComposableArchitecture'. This string must contain only alphanumeric (A-Z,a-z,0-9), and underscore (_) characters.
```

I don't get this kind of warning with my own framework, that uses similar naming convention (package name is `swift-package-name` and target name is `PackageName`)

I see that point free also uses tuist, so you guys may have some knowledge to share that I'm lacking.

**Reply by mbrandonw** [reply]
Hi @iharandreyev, this is certainly an issue with tuist and linking. We get issues like this quite frequently, but we do not personally use tuist so there isn't much we can do to help. I would suggest posting a question on their GitHub to see if they can help.

**Comment by iharandreyev** [comment]
@mbrandonw by the way, why is `perceptionRegistrar` still used when `Observation` can be imported?
I thought the idea was of `perceptionRegistrar` was to be a placeholder for `observationRegistrar` when `Observation` and `iOS 17` are not available.

**Reply by mbrandonw** [reply]
The point of `PerceptionRegistrar` is to provide a single interface for observation that works on both iOS 17 and iOS 16 (and lower). It [determines at runtime](https://github.com/pointfreeco/swift-perception/blob/9f6c437400bead83f98da0d98e129e68914b1e2d/Sources/Perception/PerceptionRegistrar.swift#L30-L39
) whether or not to call down to `ObservationRegistrar`, if available, or our back port `_PerceptionRegistrar`.

**Comment by iharandreyev** [comment]
Created [an issue](https://github.com/tuist/tuist/issues/6320) in the tuist project, hope someone helps

**Reply by mbrandonw** [reply]
Thanks for doing that!

**Reply by Narsail** [reply]
@iharandreyev How are you linking ComposableArchitecture to your project? I'm migrating my project to tuist at the moment and I ran into various linking issues as well. What works at the moment (at least with Composable Architecture, I still have other linker issues) is the following setup:

App -> Shared **Static** Framework -> **Dynamic** Composable Architecture 

That said, I had some weird crashes when I used the LDOtherFlags: "-load_all" which created the issue of having multiple libraries of Composable Architecture linked resulting in weird crashes where casting from and to the same state crashed the app. I had to use "-load_all" to solve a SwiftUI issue that I encountered with my SharedFramework.

**Reply by iharandreyev** [reply]
You can check the configuration in the [sample project repo](https://github.com/iharandreyev/Tuist-CTA-LinkingIssue).
In short: I link TCA as `.staticLibrary` to the app target using tuist package. Linking it as `.dynamicLibrary` does not help though.
So in my case, there's no intermediate `Shared Static Framework` (at least for now, I am planning to move everything TCA-related into some `Core` package)

**Reply by Narsail** [reply]
Alright, thanks for opening the issue with Tuist. I will follow it closely 🙏
